dlpy/applications/enet.py

```diff
@@ -0,0 +1,271 @@
+#!/usr/bin/env python
+# encoding: utf-8
+#
+# Copyright SAS Institute
+#
+#  Licensed under the Apache License, Version 2.0 (the License);
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+from dlpy.model import Model
+from dlpy.sequential import Sequential
+from dlpy.layers import (Conv2d, Input, Pooling, BN, Conv2DTranspose, Concat, Res, Segmentation)
+from .application_utils import get_layer_options, input_layer_options
+
+
+def initialBlock(inp):
+    '''
+    Defines the initial block of ENet
+
+    Parameters
+    ----------
+    inp : Input layer
+
+    Returns
+    -------
+    :class:`Concat`
+    '''
+    x = Conv2d(13, 3, stride=2, padding=1, act='identity', include_bias=False)(inp)
+    x = BN(act='relu')(x)
+    y = Pooling(2)(inp)
+    merge = Concat()([x, y])
+
+    return merge
+
+
+def downsamplingBNeck(x, in_depth, out_depth, projection_ratio=4):
+    '''
+    Defines the down-sampling bottleneck of ENet
+
+    Parameters
+    ----------
+    x :
+        Previous layer to this block
+    in_depth : int
+        Depth of the layer fed into this block
+    out_depth : int
+        Depth of the output layer of this block
+    projection_ratio : int, optional
+        Used to calculate the reduced_depth for intermediate convolution layers
+        Default: 4
+
+    Returns
+    -------
+    :class:`Res`
+    '''
+
+    reduced_depth = int(in_depth // projection_ratio)
+
+    conv1 = Conv2d(reduced_depth, 3, stride=2, padding=1, act='identity', include_bias=False)(x)
+    bn1 = BN(act='relu')(conv1)
+
+    conv2 = Conv2d(reduced_depth, 3, stride=1, act='identity', include_bias=False)(bn1)
+    bn2 = BN(act='relu')(conv2)
+
+    conv3 = Conv2d(out_depth, 1, stride=1, act='identity', include_bias=False)(bn2)
+    bn3 = BN(act='relu')(conv3)
+
+    pool1 = Pooling(2, stride=2)(x)
+    conv4 = Conv2d(out_depth, 1, stride=1, act='identity', include_bias=False)(pool1)
+    bn4 = BN(act='relu')(conv4)
+
+    res1 = Res()([bn3, bn4])
```

Comment:
maybe we can use "res" as variable name as we don't have multiple res layers in this function